politeiawww/dcc.go

```diff
@@ -0,0 +1,553 @@
+// backendDCCMetadata.
+func decodeBackendDCCMetadata(payload []byte) (*backendDCCMetadata, error) {
+	var md backendDCCMetadata
+
+	err := json.Unmarshal(payload, &md)
+	if err != nil {
+		return nil, err
+	}
+
+	return &md, nil
+}
+
+// encodeBackendDCCStatusChange encodes a backendDCCStatusChange into a
+// JSON byte slice.
+func encodeBackendDCCStatusChange(md backendDCCStatusChange) ([]byte, error) {
+	b, err := json.Marshal(md)
+	if err != nil {
+		return nil, err
+	}
+
+	return b, nil
+}
+
+// decodeBackendDCCStatusChanges decodes a JSON byte slice into a slice of
+// backendDCCStatusChanges.
+func decodeBackendDCCStatusChanges(payload []byte) ([]backendDCCStatusChange, error) {
+	var md []backendDCCStatusChange
+
+	d := json.NewDecoder(strings.NewReader(string(payload)))
+	for {
+		var m backendDCCStatusChange
+		err := d.Decode(&m)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+
+		md = append(md, m)
+	}
+
+	return md, nil
+}
+
+// getDCC gets the most recent verions of the given DCC from the cache
+// then fills in any missing user fields before returning the DCC record.
+func (p *politeiawww) getDCC(token string) (*cms.DCCRecord, error) {
+	// Get invoice from cache
+	r, err := p.cache.Record(token)
+	if err != nil {
+		return nil, err
+	}
+	i := convertDCCFromCache(*r)
+
+	// Fill in userID and username fields
+	u, err := p.db.UserGetByPubKey(i.PublicKey)
+	if err != nil {
+		log.Errorf("getDCC: getUserByPubKey: token:%v "+
+			"pubKey:%v err:%v", token, i.PublicKey, err)
+	} else {
+		i.SponsorUserID = u.ID.String()
+		i.SponsorUsername = u.Username
+	}
+	support, oppose, err := p.getDCCSupportOppositionComments(token)
+	if err != nil {
+		log.Errorf("getDCC: %v", err)
+	}
+	i.SupportUserIDs = support
+	i.OppositionUserIDs = oppose
+	return &i, nil
+}
+
+func (p *politeiawww) getDCCSupportOppositionComments(token string) ([]string, []string, error) {
+	log.Tracef("getDCCSupportOpposition: %v", token)
+
+	dc, err := p.decredGetComments(token)
+	if err != nil {
+		return nil, nil, fmt.Errorf("decredGetComments: %v", err)
+	}
+
+	support := make([]string, 0, len(dc))
+	oppose := make([]string, 0, len(dc))
+	for _, v := range dc {
+		c := convertCommentFromDecred(v)
+		u, err := p.db.UserGetByPubKey(c.PublicKey)
+		if err != nil {
+			log.Errorf("getDCCSupportOpposition: UserGetByPubKey: "+
+				"token:%v commentID:%v pubKey:%v err:%v",
+				token, c.CommentID, c.PublicKey, err)
+		}
+		if c.Comment == sponsorString {
+			support = append(support, u.ID.String())
+		} else if c.Comment == opposeString {
+			oppose = append(support, u.ID.String())
+		}
+	}
+
+	return support, oppose, nil
+}
+
+func (p *politeiawww) processDCCDetails(gd cms.DCCDetails) (*cms.DCCDetailsReply, error) {
+	log.Tracef("processDCCDetails: %v", gd.Token)
+
+	dcc, err := p.getDCC(gd.Token)
+	if err != nil {
+		if err == cache.ErrRecordNotFound {
+			err = www.UserError{
+				ErrorCode: cms.ErrorStatusInvoiceNotFound,
```

Comment:
A DCC is a separate type of document than an invoice.  Should this be a DCC specific error?